File: Gramaticas/Ejercicio5/C/ejercicio5.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>

// L(G5) = { a (ab)^n b | n >= 0 }
bool es_ejercicio5(char *s) {
    int n = strlen(s);

    if (n < 2) return false;

    if (s[0] != 'a' || s[n-1] != 'b') {
        return false;
    }

    int i = 1;
    while (i < n-1) {
        if (s[i] == 'a' && s[i+1] == 'b') {
            i += 2;
        } else {
            return false;
        }
    }

    return i == n-1;
}

int main() {
    FILE *f = fopen("ejercicio5.txt", "r");
    if (!f) {
        printf("Error al abrir ejercicio5.txt\n");
        return 1;
    }

    char linea[256];
    while (fgets(linea, sizeof(linea), f)) {
        linea[strcspn(linea, "\n")] = 0;
        if (es_ejercicio5(linea)) printf("%s => ACEPTA\n", linea);
        else printf("%s => NO ACEPTA\n", linea);
    }
    fclose(f);
    return 0;
}

File: Gramaticas/Ejercicio5/Python/ejercicio5.py
# L(G5) = { a (ab)^n b | n >= 0 }

def es_ejercicio5(cadena):
    if len(cadena) < 2:
        return False

    # Debe empezar en 'a' y terminar en 'b'
    if not (cadena[0] == "a" and cadena[-1] == "b"):
        return False

    # Quitamos la primera y la última letra
    medio = cadena[1:-1]

    # Revisamos que esté compuesto únicamente por repeticiones de "ab"
    i = 0
    while i < len(medio):
        if i + 1 < len(medio) and medio[i] == "a" and medio[i+1] == "b":
            i += 2
        else:
            return False

    return True

def procesar(archivo):
    with open(archivo, "r") as f:
        for linea in f:
            cad = linea.strip()
            if es_ejercicio5(cad):
                print(cad, "=> ACEPTA")
            else:
                print(cad, "=> NO ACEPTA")

if __name__ == "__main__":
    procesar("ejercicio5.txt")
